atsign-foundation/at_python
# feat: Notification monitoring system

**- What I did**
Implement AtMonitorConnection and notification monitoring system.

**- How I did it**
- Took [AtMonitorConnection.java](https://github.com/atsign-foundation/at_java/blob/trunk/at_client/src/main/java/org/atsign/client/api/impl/connections/AtMonitorConnection.java) as an example and implemented it, but used `locks` instead of  `synchronized` (because synchronized doesn't exist in Python, it needs to be created as a decorator).

- Created `AtEventType` enum and implemented an event system based on [Queue](https://docs.python.org/3/library/queue.html) library. There were EventBus solutions based on `asyncio` lib, but I didn't find practical the added complexity of mixing `Threading` and `asyncio` for now.

- Created `AtEvent` class, which is essentially a wrapper for `event + data associated to the event`. `AtEvent` objects are then introduced into the `Queue` and read by other thread which will use that data.

- Added new parameters which will be eventually necessary to `Metadata` class (although they are commented because there's a circular import issue).

- `TimeUtil` to manage milliseconds, `SocketUtil` to create `readline()` method for `SSLSocket`.

- Created `AtNotification` class, although it isn't used for now.

- Added `start_monitor` and `stop_monitor` methods to `AtClient` class, which can be used to start the monitoring for notifications. Added `handle_event` method to `AtClient` class.

- `atconstants.py` is simply a file to define constants or global variables, such as locks or queues.

- Updated `repl.py`, which now includes a practical example of how this new feature can be used (simply create a new thread which calls a method, and handle the event reception through queues inside that method, also calling inside of it the `handle_event` method of `AtClient` class).

- Small tweaks not mentioned, feel free to ask.

- Issues: 
1. `Metadata` circular imports exception if new code is uncommented.
2. Unexpected `MONITOR_EXCEPTION` events triggered when reading `@atSign@notification` notifications from yourself (doesn't happen with normal `notification` notifications also from yourself, which are actually the same as the other mentioned).
3. Current way to manage events with `handle_event` methods could be improved to a more abstract structure, although they work for now.

**- How to verify it**

1. Run `repl.py` and choose an `atSign` using option `1`
2. Select option `2`. REPL will start and activate monitor mode automatically in a different thread. You can still send commands/verbs. You will start seeing your own notifications (from yourself to yourself) and heartbeat working (noop verb is sent from time to time as a keepalive)
3. Use `at_talk` or any other tool to send notifications to your `atSign` from a different `atSign`. You should be able to see the complete notification, and the encrypted and decrypted value of it.

**- Description for the changelog**

feat: Notification monitoring system

Progress in https://github.com/atsign-foundation/at_python/issues/8 and https://github.com/atsign-foundation/at_python/issues/9
